include/net/nrf_cloud_logs.h

```diff
@@ -0,0 +1,98 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef NRF_CLOUD_LOGS_H_
+#define NRF_CLOUD_LOGS_H_
+
+/** @file nrf_cloud_logs.h
+ * @brief Module to provide nRF Cloud logging support to nRF9160 SiP.
+ */
+
+#include <zephyr/logging/log.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct nrf_cloud_rest_context;
+
+/** @defgroup nrf_cloud_logs nRF Cloud Logs
+ * @{
+ */
+
+/** @brief Initialize the nRF Cloud logging subsystem. Call after
+ *  the date_time library is ready.
+ */
+void nrf_cloud_log_init(void);
+
+/** @brief Set log_level to the level at or below which a log message will be
+ *  sent to the cloud. Log messages with a log level larger than this
+ *  will not be sent. For example, if log_level is LOG_LEVEL_WRN (2), then
+ *  LOG_LEVEL_WRN (2) and LOG_LEVEL_ERR (1) messages will be sent, and
+ *  LOG_LEVEL_INF (3) and LOG_LEVEL_DBG (4) will not.
+ *  Set log_level to LOG_LEVEL_NONE (0) to disable logging.
+ *  See zephyr/include/zephyr/logging/log_core.h.
+ *
+ * @param log_level The log level to go to the cloud.
+ */
+void nrf_cloud_log_control_set(int log_level);
+
+/** @brief Get current log level to be sent to nRF Cloud.
+ * @retval Current log level, if CONFIG_NRF_CLOUD_LOG_BACKEND is enabled, otherwise
+ *         0, indicating logs are disabled.
+ */
+int nrf_cloud_log_control_get(void);
+
+/** @brief Tell REST-based logger the REST context and device_id for later
+ *         use when outputting a log.
+ *
+ *  @param[in] ctx REST context to use.
+ *  @param[in] dev_id Device ID to send message on behalf of.
+ */
+void nrf_cloud_rest_log_context_set(struct nrf_cloud_rest_context *ctx, const char *dev_id);
+
+/** @brief Enable or disable logging to the cloud.
+ *  @param[in] enable Set true to send logs to the cloud, false to disable.
+ */
+void nrf_cloud_log_enable(bool enable);
+
+/** @brief Find out if logger is enabled.
+ *
+ *  @retval True if logger is enabled.
+ */
+bool nrf_cloud_log_is_enabled(void);
+
+#if defined(CONFIG_NRF_CLOUD_MQTT)
+/** @brief Directly log to the cloud. This does not use the Zephyr logging
+ *  system if CONFIG_NRF_CLOUD_LOG_BACKEND is disabled.  Otherwise, it is passed to
+ *  the logging system.
+ *
+ *  @param log_level The log level for this message; dropped if above set log level.
+ *  @param fmt Printf format specifier.
+ */
+int nrf_cloud_log_send(int log_level, const char *fmt, ...);
+
+#elif defined(CONFIG_NRF_CLOUD_REST)
+/** @brief Directly log to the cloud. This does not use the Zephyr logging
+ *  system if CONFIG_NRF_CLOUD_LOG_BACKEND is disabled.  Otherwise, it is passed to
+ *  the logging system.
+ *
+ *  @param ctx REST interface context to use to send log message.
+ *  @param dev_id Device id to use when sending the log message via REST.
+ *  @param log_level The log level for this message; dropped if above set log level.
+ *  @param fmt Printf format specifier.
+ */
+int nrf_cloud_rest_log_send(struct nrf_cloud_rest_context *ctx, const char *dev_id,
```

Comment:
I was following @jayteemo's request:
https://github.com/nrfconnect/sdk-nrf/pull/10266#pullrequestreview-1322686250

Justin, what do you think of @tokangas's suggestion?

Reply:
REST and MQTT are quite different.
A user might want the ability to control when their device initiates the https connection for REST.
For example, the device wakes up, connects to the network, initiates the https socket connection, makes some REST calls, disconnects socket connection, goes back to sleep.

This also aligns with the other nrf_cloud_rest functions.

Reply:
Ok, if there's a need to have it separate, it's fine. I see that alerts has a similar REST specific function, but just wondering if those should actually be in `nrf_cloud_rest.h` instead to be consistent. Everything else REST specific is in there.

Reply:
> I see that alerts has a similar REST specific function, but just wondering if those should actually be in `nrf_cloud_rest.h` instead to be consistent. Everything else REST specific is in there.

I started to think about this too. Without having an opinion on whether `nrf_cloud_rest_*()` functions should be in the modules where the corresponding functionality is (`nrf_cloud_agps`, `nrf_cloud_location`, etc.), given it has been selected to put REST specific functions into `nrf_cloud_rest`, it would be logical to put the log functions there too.
What to do with alerts that doesn't follow this principle is another question.
Or is there a specific reason why alerts and logs shouldn't follow the principle made with agps, location, etc?

Reply:
I chose to do the alerts and logs this way (self-contained) is that the REST-specific involvement is quite minimal.  I use a single rest data type:`struct nrf_cloud_rest_context`, and make a single REST function call: `nrf_cloud_rest_send_device_message()`.  This allows me to keep like with like, separate concerns, and not spread these features over multiple different files.

Reply:
I understand having alerts and logs as self-contained is nice, but consistency with other modules done earlier would be nice too. If you decide to go to self-contained logs with REST and MQTT in the same file, you break the consistency and it has some small inconsistent implications too. Like you cannot name the function as other rest functions (`nrf_cloud_rest_log_send()`) if you keep the naming based on the filename (`nrf_cloud_log_rest_send()`).

But when there's a tradeoff to be made, I often vote for consistency.